modules/dag/src/main/scala/org/tesselation/dag/l1/DAGBlockValidator.scala

```diff
@@ -0,0 +1,170 @@
+      )
+      _ <- validateBlock(signedBlock.value).ifM(
+        Async[F].unit,
+        Async[F].raiseError[Unit](new Throwable("Invalid block data!"))
+      )
+      hashedBlock <- signedBlock.hashWithSignatureCheck.flatMap(_.liftTo[F])
+
+      hashedTransactions <- signedBlock.value.transactions.traverse(_.hashWithSignatureCheck.flatMap(_.liftTo[F]))
+
+      _ <- acceptTransactions(hashedTransactions)
+      _ <- blockStorage.acceptBlock(hashedBlock)
+    } yield ()
+
+  def validateBlockSignatures(signedBlock: Signed[DAGBlock]): F[Boolean] =
+    for {
+      areSignaturesValid <- signedBlock.hasValidSignature
+      areAtLeastThreeSignatures = signedBlock.proofs.map(_.id).toList.toSet.size >= 3 //At least 3 distinct signers
+    } yield areSignaturesValid && areAtLeastThreeSignatures
+
+  def validateBlock(block: DAGBlock): F[Boolean] =
+    for {
+      areParentsAccepted <- areParentsAccepted(block)
+      areTxsValid <- areTransactionsValid(block.transactions)
+    } yield areParentsAccepted && areTxsValid
+
+  private def areParentsAccepted(block: DAGBlock): F[Boolean] =
+    blockStorage.areParentsAccepted(block)
+
+  // TODO: for now I think last transaction stored is enough, I'm not storing all the transactions in transaction
+  //       service as it was the case in constellation - all in all, we will store blocks and redownload based on them
+  //       and anyway the check for transactions forming a correct chain should catch this case of accepted or stale transactions trying to pass validation
+//  def containsAcceptedOrStaleTx(cbd: DAGBlockData): F[List[String]] = {
+//    val containsAccepted = cbd.transactions.toList.map { t =>
+//
+//      transactionStorage.getLastAcceptedTransactionRef(t.baseHash).map {
+//        //b => (t.hash, b)
+//        case Some(tr) =>
+//        case None =>
+//      }
+//    }.sequence[F, (String, Boolean)]
+//      .map(l => l.collect { case (h, true) => h })
+//
+//    containsAccepted
+//  }
+
+  private def areTransactionsValid(signedTransactions: Seq[Signed[Transaction]]): F[Boolean] =
+    for {
+      arePassingValidation <- areTransactionsPassingValidation(signedTransactions)
+      areChainedCorrectly <- areAllTransactionsFormingCorrectChain(signedTransactions.map(_.value))
+    } yield arePassingValidation && areChainedCorrectly
+
+  private def areTransactionsPassingValidation(signedTransactions: Seq[Signed[Transaction]]): F[Boolean] =
+    transactionValidator
+      .validateBatchOfTransactions(signedTransactions)
+      .map(_.isValid)
+      .handleError(_ => false)
+
+  private def areAllTransactionsFormingCorrectChain(transactions: Seq[Transaction]): F[Boolean] =
+    transactions
+      .groupBy(_.source)
+      .toList
+      .traverse {
+        case (address, txs) =>
+          for {
+            lastAccepted <- transactionStorage.getLastAcceptedTransactionRef(address)
+            sorted = txs.sortBy(_.parent.ordinal.coerce.value)
```

Comment:
you should implement Order instance for ordinal, and then just sort by ordinal

Reply:
Right, I actually have it defined, just need to use it.